File: webapp/src/Websocket/types/UserSettings.ts
export type UserRole =
    "garageDoor"
    | "evCharging"
    | "energyStorageSystem"
    | "energyPricing"
    | "heaterUnderFloor"
    | "heaterStairs"
    | "environmentSensors"
    | "cameras"
    | "recordings"
    | "userSettings"
    | "firewall"
    ;

export type Level = "none" | "read" | "readWrite" | "readWriteAdmin";

export type UserSettings = {
    authorization: AuthorizationMap;
}

export type AuthorizationMap = {
    [key in UserRole]: Level;
}

export type WriteCommand = "addUser" | "removeUser" | "updateAuthorization";

export type WriteUserSettings = {
    user: string;
    command: WriteCommand;
    writeAuthorization?: WriteAuthorization
}

export type WriteAuthorization = {
    userRole: UserRole,
    level: Level,
}

File: src/main/kotlin/com/dehnes/smarthome/api/dtos/UserSettings.kt
package com.dehnes.smarthome.api.dtos

import com.dehnes.smarthome.users.Level
import com.dehnes.smarthome.users.UserRole


data class WriteUserSettings(
    val user: String,
    val command: WriteCommand,
    val writeAuthorization: WriteAuthorization?
)

enum class WriteCommand {
    addUser,
    removeUser,
    updateAuthorization
}

data class WriteAuthorization(
    val userRole: UserRole,
    val level: Level,
)
